debug_scripts/estimate_epoch_start_time.py

```diff
@@ -0,0 +1,142 @@
+    )
+
+    return epoch_lengths, exponential_weighted_average_epoch_length
+
+
+# Function to approximate future epoch start dates
+def predict_future_epochs(starting_epoch_timestamp, avg_epoch_length,
+                          num_future_epochs):
+    future_epochs = []
+    current_timestamp = starting_epoch_timestamp / 1e9  # Convert from nanoseconds to seconds
+
+    for i in range(1, num_future_epochs + 1):
+        # Add the average epoch length for each future epoch
+        future_timestamp = current_timestamp + (i * avg_epoch_length)
+
+        # Convert to human-readable format
+        future_date = time.strftime('%Y-%m-%d %H:%M:%S %A',
+                                    time.gmtime(future_timestamp))
+        future_epochs.append(future_date)
+
+        print(f"Predicted start of epoch {i}: {future_date}")
+
+    return future_epochs
+
+
+# First, we get the latest block and save the next_epoch_id
+def get_latest_block(url):
+    payload = {
+        "jsonrpc": "2.0",
+        "id": "dontcare",
+        "method": "block",
+        "params": {
+            "finality": "final"
+        }
+    }
+    response = requests.post(url, json=payload).json()
+    return response['result']['header']
+
+
+# Main function to run the process
+def main(args):
+    latest_block = get_latest_block(args.url)
+    next_epoch_id = latest_block['next_epoch_id']
+    current_epoch_first_block = get_block_by_hash(args.url, next_epoch_id)
+    current_timestamp = int(current_epoch_first_block['timestamp']
+                           )  # Current epoch start timestamp in nanoseconds
+
+    # Get epoch lengths and the exponential weighted average
+    epoch_lengths, exponential_weighted_average_epoch_length = get_exponential_weighted_epoch_lengths(
+        args.url, next_epoch_id, args.num_epochs, args.decay_rate)
+
+    # Predict future epoch start dates
+    predict_future_epochs(current_timestamp,
+                          exponential_weighted_average_epoch_length,
+                          args.num_future_epochs)
+
+
+# Set up command-line argument parsing
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Approximate future epoch start dates for NEAR Protocol.")
+    parser.add_argument("--url", required=True, help="The RPC URL to query.")
+    parser.add_argument("--num_epochs",
+                        type=int,
+                        default=4,
+                        help="Number of past epochs to analyze.")
+    parser.add_argument("--decay_rate",
+                        type=float,
+                        default=0.1,
+                        help="Decay rate for exponential weighting.")
+    parser.add_argument("--num_future_epochs",
+                        type=int,
+                        default=3,
+                        help="Number of future epochs to predict.")
+
+    args = parser.parse_args()
+    main(args)
```

Comment:
Very useful, thank you for making this!

To resolve https://github.com/near/nearcore/issues/12107 I would love to have two more things:
* If the voting date is X, when will the protocol upgrade happen?
* Given a day of the voting date find the hour that maximizes the probability that the upgrade happens during working hours

But that can be done in follow up PRs